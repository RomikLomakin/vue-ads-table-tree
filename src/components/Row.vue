<template>
    <tr
        :class="rowClasses"
    >
        <vue-ads-cell
            v-for="(column, key) in columns.items"
            slot="slot(column)"
            :key="key"
            :row-index="$vnode.key"
            :row="row"
            :column="column"
            :classes="classes"
            @toggleChildren="$emit('toggleChildren');"
        />
    </tr>
</template>

<script>
import Row from '../models/Row';
import VueAdsCell from './Cell';

import ClassProcessor from '../services/ClassProcessor';

export default {
    name: 'VueAdsRow',

    components: {
        VueAdsCell,
    },

    props: {
        row: {
            type: Row,
            required: true,
        },

        columns: {
            type: Object,
            required: true,
        },

        slots: {
            type: Object,
            required: false,
            default: () => { return {}; },
        },

        classes: {
            type: ClassProcessor,
            required: true,
        },
    },

    computed: {
        rowClasses () {
            return Object.assign(
                this.classes.process(this.$vnode.key + 1, null, this.row),
                this.row.classes ? ClassProcessor.processValue(this.row.classes.row, this.row) : {}
            );
        },
    },

    methods: {
        columnSlot (column) {
            return this.slots[column.property + '_' + this.row[column.property]] ||
                this.slots[column.property] ||
                null;
        },
    },
};
</script>
