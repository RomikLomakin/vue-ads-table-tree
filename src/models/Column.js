export default class Column {
    constructor (properties) {
        Object.assign(this, properties);
    }

    set property (property) {
        this._property = property;
    }

    get property () {
        if (!this.hasOwnProperty('_property')) {
            this._property = '';
        }

        return this._property;
    }

    set order (order) {
        this._order = order;
    }

    get order () {
        if (!this.hasOwnProperty('_order')) {
            return 0;
        }

        return this._order;
    }

    set collapseIcon (collapseIcon) {
        this._collapseIcon = collapseIcon;
    }

    get collapseIcon () {
        if (!this.hasOwnProperty('_collapseIcon')) {
            this._collapseIcon = false;
        }

        return this._collapseIcon;
    }

    set sortable (sortable) {
        this._sortable = sortable;
    }

    get sortable () {
        if (!this.hasOwnProperty('_sortable')) {
            this._sortable = false;
        }

        return this._sortable;
    }

    set direction (direction) {
        if (![true, false, null].includes(direction)) {
            throw new Error(
                'Sort direction has to be true (asc), false (desc) or null (no sorting)',
            );
        }

        this._direction = direction;
    }

    get direction () {
        if (!this.hasOwnProperty('_direction')) {
            this._direction = null;
        }

        return this._direction;
    }

    set filterable (filterable) {
        this._filterable = filterable;
    }

    get filterable () {
        if (!this.hasOwnProperty('_filterable')) {
            this._filterable = false;
        }

        return this._filterable;
    }

    sort (order) {
        this.direction = !this.direction;
        this.order = order;
    }
}
