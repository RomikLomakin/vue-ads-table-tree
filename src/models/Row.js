import RowCollection from '../collections/RowCollection';

export default class Row {
    constructor (properties) {
        this.children = new RowCollection();
        this.processedChildren = null;
        this.showChildren = false;
        this.hasChildren = false;
        this.childrenLoading = false;
        Object.assign(this, properties);
    }

    set parent (parent) {
        if (!(parent instanceof Row)) {
            throw new Error(
                'Parent of a row has to be a row'
            );
        }

        this._parent = parent;
    }

    get parent () {
        return this._parent;
    }

    set children (children) {
        this._children = children instanceof RowCollection ? children : new RowCollection(children);
        this._children.items
            .forEach(childRow => {
                childRow.parent = this;
            });

        this.hasChildren = !this._children.isEmpty();
    }

    get children () {
        return this._children;
    }

    set childrenLoading (childrenLoading) {
        this._childrenLoading = childrenLoading;
    }

    get childrenLoading () {
        return this._childrenLoading;
    }

    set processedChildren (processedChildren) {
        this._processedChildren = processedChildren;

        // wss voor async calls.
        // this._processedChildren.items
        //     .forEach(filteredChild => {
        //         filteredChild.parent = this;
        //     });
    }

    get processedChildren () {
        return this._processedChildren === null ? this.children : this._processedChildren;
    }

    set hasChildren (hasChildren) {
        this._hasChildren = hasChildren;
    }

    get hasChildren () {
        return this._hasChildren;
    }

    set showChildren (showChildren) {
        this._showChildren = showChildren;
    }

    get showChildren () {
        return this._showChildren;
    }

    get properties () {
        return Object.getOwnPropertyNames(this)
            .filter(property => {
                return property[0] !== '_';
            });
    }

    toggleChildren () {
        this.showChildren = !this.showChildren;
    }

    childrenLoaded () {
        return this.hasChildren && !this.children.isEmpty();
    }

    loadChildren () {
        return this.showChildren && this.hasChildren && this.children.isEmpty();
    }

    countParents () {
        if (this.parent) {
            return this.parent.countParents() + 1;
        }

        return 0;
    }
}
