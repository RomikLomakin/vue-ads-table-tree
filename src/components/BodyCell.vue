<script>
import Border from '../models/Border';
import Row from '../models/Row';
import Column from '../models/Column';
import ToggleChildrenButton from './ToggleChildrenButton';

export default {
    name: 'BodyCell',

    components: {
        ToggleChildrenButton,
    },

    props: {
        border: {
            type: Border,
            required: true,
        },

        row: {
            type: Row,
            required: true,
        },

        column: {
            type: Column,
            required: true,
        },

        index: {
            type: Number,
            required: true,
        },
    },

    data () {
        return {
            // loading: false,
        };
    },

    render (createElement) {
        return createElement('td', {
            class: this.cellClasses,
            style: this.cellStyle,
        }, [
            createElement('span', {}, this.value(createElement)),
        ]);
    },

    computed: {
        cellClasses () {
            return Object.assign(
                this.border.columnClasses(this.column.last),
                {
                    'px-4': true,
                    'py-2': true,
                    'text-sm': true,
                    ['w-' + this.column.width]: true,
                }
            );
        },

        cellStyle () {
            let paddingLeft = 1 + this.first * this.row.countParents() * 1.5;

            return {
                'padding-left': paddingLeft + 'rem',
            };
        },

        first () {
            return this.index === 0;
        },

        // isLoading () {
        //     return this.loading && this.row.children.isEmpty() && this.row.processedChildren.isEmpty();
        // },
    },

    methods: {
        value (createElement) {
            let elements = [];

            if (this.first && this.row.hasChildren) {
                elements.push(
                    createElement(ToggleChildrenButton, {
                        props: {
                            expanded: this.row.showChildren,
                            // loading: this.loading,
                        },
                        nativeOn: {
                            'click': this.toggleChildren,
                        },
                    }),
                );
            }

            if (this.column.property && this.row.hasOwnProperty(this.column.property)) {
                elements.push(this.row[this.column.property]);
            }

            return elements.length > 0 ? elements : [''];
        },

        toggleChildren () {
            // this.loading = this.row.hasLoadedChildren();
            this.$emit('toggleChildren');
        },
    },
};
</script>
