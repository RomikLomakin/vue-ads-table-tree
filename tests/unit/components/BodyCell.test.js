import { shallowMount } from '@vue/test-utils';

import BodyCell from '../../../src/components/Cell';
import Row from '../../../src/models/Row';
import Column from '../../../src/models/Column';
import Styling from '../../../src/models/Styling';

describe('BodyCell', () => {
    let bodyCell;

    beforeEach(() => {
        bodyCell = shallowMount(BodyCell, {
            propsData: {
                styling: new Styling({
                    columnsAllExceptLast: 'border-r',
                }),
                row: new Row({
                    firstName: 'arne',
                }),
                column: new Column({
                    property: 'firstName',
                }),
                index: 0,
            },
        });
    });

    it('returns the default cell classes', () => {
        expect(bodyCell.vm.classes).toEqual({
            'border-r': true,
            'px-4': true,
            'py-2': true,
            'text-sm': true,
            'w-auto': true,
        });

        expect(bodyCell.vm.first).toBeTruthy();
        expect(bodyCell.vm.style['padding-left']).toBe('1rem');

        expect(bodyCell.html()).toMatchSnapshot();
    });

    it('changes the padding if the number of parents changes', () => {
        bodyCell.setProps({
            row: new Row({
                parent: new Row(),
            }),
        });

        expect(bodyCell.vm.style['padding-left']).toBe('2.5rem');
    });

    it('adds a toggle children button if the row has children', () => {
        bodyCell.setProps({
            row: new Row({
                hasChildren: true,
            }),
        });

        expect(bodyCell.html()).toMatchSnapshot();
    });

    it('is empty with no matching properties', () => {
        bodyCell.setProps({
            row: new Row({
                lastName: 'de smedt',
            }),
        });
    });

    it('emits toggle children, when calling toggle children', () => {
        bodyCell.vm.toggleChildren();

        expect(bodyCell.emitted().toggleChildren).toBeTruthy();
    });
});
