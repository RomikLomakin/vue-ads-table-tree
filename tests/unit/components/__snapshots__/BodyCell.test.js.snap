// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BodyCell adds a toggle children button if the row has children 1`] = `<td class="border-r  px-4 py-2 text-sm w-auto" style="padding-left: 1rem;"><span><i class="fa pr-2 cursor-pointer fa-plus-square"></i></span></td>`;

exports[`BodyCell returns the default cell classes 1`] = `<td class="border-r  px-4 py-2 text-sm w-auto" style="padding-left: 1rem;"><span>arne</span></td>`;
